gremlin-go/driver/resultSet.go

```diff
@@ -50,25 +50,60 @@ type channelResultSet struct {
 	statusAttributes map[string]interface{}
 	closed           bool
 	err              error
-	mux              sync.Mutex
+	waitSignal       chan bool
+	channelMux       sync.Mutex
+	waitSignalMux    sync.Mutex
+}
+
+func (channelResultSet *channelResultSet) sendSignal() {
+	// Lock wait
+	channelResultSet.waitSignalMux.Lock()
+	if channelResultSet.waitSignal != nil {
+		channelResultSet.waitSignal <- true
+		channelResultSet.waitSignal = nil
+	}
+	channelResultSet.waitSignalMux.Unlock()
 }
 
 func (channelResultSet *channelResultSet) GetError() error {
 	return channelResultSet.err
 }
 
 func (channelResultSet *channelResultSet) IsEmpty() bool {
-	channelResultSet.mux.Lock()
-	defer channelResultSet.mux.Unlock()
-	return channelResultSet.closed && len(channelResultSet.channel) == 0
+	channelResultSet.channelMux.Lock()
+	// If our channel is empty and we have no data in it, wait for signal that the state has been updated.
+	channelIsEmpty := len(channelResultSet.channel) == 0
+	if !channelIsEmpty || channelResultSet.closed {
```

Comment:
Should it not return True if channel is not empty and we have the channel lock in possession at this point?

Reply:
Not necessarily. This logic here is a bit more hard to digest as Simon said, I was using some boolean logic equivalence to reduce my flow control but I am going to make it more straightforward instead.